src/client/metric.rs

```diff
@@ -286,11 +291,170 @@ impl<T: MetricType + Clone> Metric<T> {
     }
 }
 
+pub struct InstanceMetric<T> {
+    pub (super) metrics: HashMap<String, Metric<T>>,
+    sem: Semantics,
+    unit: Unit,
+    indom: u32,
+    shorthelp: String,
+    longhelp: String,
+}
+
+impl<T: MetricType + Clone> InstanceMetric<T> {
+    pub fn new(
+        sem: Semantics,
+        unit: Unit,
+        shorthelp_text: &str,
+        longhelp_text: &str) -> Result<InstanceMetricBuilder<T>, String> {
+        InstanceMetricBuilder::new(sem, unit, shorthelp_text, longhelp_text)
+    }
+
+    pub fn indom(&self) -> u32 {
+        self.indom
+    }
+
+    pub fn instance_count(&self) -> u32 {
+        self.metrics.len() as u32
+    }
+
+    pub fn has_instance(&self, instance: &str) -> bool {
+        self.metrics.contains_key(instance)
+    }
+
+    pub fn val(&self, instance: &str) -> Option<T> {
+        self.metrics.get(instance).map(|m| m.val())
+    }
+
+    pub fn set_val(&mut self, instance: &str, val: T) -> Option<io::Result<()>>  {
+        self.metrics.get_mut(instance)
+            .map(|m| m.set_val(val))
+    }
+
+    pub fn shorthelp(&self) -> &str { &self.shorthelp }
+    pub fn longhelp(&self) -> &str { &self.longhelp }
+
+    pub (super) fn instance_id(instance: &str) -> u32 {
+        let mut hasher = DefaultHasher::new();
+        hasher.write(instance.as_bytes());
+        hasher.finish() as u32
+    }
+}
+
+pub struct InstanceMetricBuilder<T> {
+    instance_metric: InstanceMetric<T>,
+    running_hasher: DefaultHasher
+}
+
+impl<T: MetricType + Clone> InstanceMetricBuilder<T> {
+    pub fn new(
+        sem: Semantics,
+        unit: Unit,
+        shorthelp_text: &str,
+        longhelp_text: &str) -> Result<Self, String>  {
+
+        if shorthelp_text.len() >= STRING_BLOCK_LEN as usize {
+            return Err(format!("short help text longer than {} bytes", STRING_BLOCK_LEN - 1));
+        }
+        if longhelp_text.len() >= STRING_BLOCK_LEN as usize {
+            return Err(format!("long help text longer than {} bytes", STRING_BLOCK_LEN - 1));
+        }
+
+        Ok(InstanceMetricBuilder {
+            instance_metric: InstanceMetric {
+                metrics: HashMap::new(),
+                sem: sem,
+                unit: unit,
+                shorthelp: shorthelp_text.to_owned(),
+                longhelp: longhelp_text.to_owned(),
+                indom: 0
+            },
+            running_hasher: DefaultHasher::new()
+        })
+    }
+
+    pub fn instance(
+        mut self,
+        instance: &str,
+        init_val: T,
+        shorthelp_text: &str,
+        longhelp_text: &str) -> Result<Self, String> {
+
+        { // this block ensures that `im` goes out of scope before we return Ok(self)
+
+            let mut im = &mut self.instance_metric;
+            instance.hash(&mut self.running_hasher);
+            im.indom = (self.running_hasher.finish() as u32) & ((1 << INDOM_BIT_LEN) - 1);
+
+            let metric = Metric::new(
+                &instance,
```

Comment:
I believe the idea here is that the Instance Metric contains inside it Singleton Metrics that have the same name as the instance name. This way, we can maintain unique name, shorthelp and longhelp, but duplicate semantics, unit and indom for each instance